### PR TITLE
E2E: Use Kyverno to enforce security constraints

### PR DESCRIPTION
**Pull request by barkbay**
Alternative to https://github.com/elastic/cloud-on-k8s/pull/5820/

I think I'll close #5820 in favour of this one as [Kyverno](https://kyverno.io/) is simpler to apprehend. The Open Policy Agent is great, the idea of having `ConstraintTemplates` is interesting at a large scale, when there's a lot of projects on a cluster. It might however be a bit too involved wrt what we are trying to achieve here. We only want to enforce basic "static" policies on a couple of namespaces, and I don't think we really need the full power of the Rego policy language. 

Fixes https://github.com/elastic/cloud-on-k8s/issues/5726

### CODE REVIEW

**Comment by barkbay** [comment]
run/e2e-tests tags=agent

**Comment by barkbay** [comment]
run/e2e-tests tags=agent

**Comment by barkbay** [comment]
run/e2e-tests

**Review by pebrc — APPROVED** [review]

File: hack/deployer/runner/kyverno/kyverno.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+var policies string
+
+const (
+	waitForKyvernoDeployment = `wait deployment kyverno -n kyverno --for condition=Available=True --timeout=60s`
+)
+
+func Install(globalKubectlOptions ...string) error {
+	k := NewKubectl(globalKubectlOptions...)
+	// Kyverno related manifests are stored in a temporary directory
+	dir, err := ioutil.TempDir("", "gatekeeper")
```
Comment (thbkrkr) [inline hack/deployer/runner/kyverno/kyverno.go:32]: Nit:
```suggestion
	dir, err := ioutil.TempDir("", "kyverno")
```

File: hack/deployer/runner/kyverno/kyverno.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(dir)
+
+	// Install Gatekeeper
```
Comment (thbkrkr) [inline hack/deployer/runner/kyverno/kyverno.go:38]: `log.Println("Installing Kyverno")` below is enough?
```suggestion
```

File: hack/deployer/runner/kyverno/kyverno.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/elastic/cloud-on-k8s/v2/hack/deployer/exec"
+)
+
+//go:embed install/kyverno.yaml
+var install string
```
Comment (thbkrkr) [inline hack/deployer/runner/kyverno/kyverno.go:20]:
```suggestion
var installerManifest string
```

File: hack/deployer/runner/kyverno/kyverno.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+//go:embed install/policies.yaml
+var policies string
```
Comment (thbkrkr) [inline hack/deployer/runner/kyverno/kyverno.go:23]:
```suggestion
var policiesManifest string
```

File: hack/deployer/runner/kyverno/kyverno.go (hunk 1 above)
Comment (thbkrkr) [reply hack/deployer/runner/kyverno/kyverno.go:32]: - #5927:
```suggestion
	dir, err := os.MkdirTemp("", "kyverno")
```

File: hack/deployer/runner/kyverno/kyverno.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		return err
+	}
+
+	log.Println("Kyverno successfully installed")
+	return nil
+}
+
+func apply(k *Kubectl, workDir string, content string, tmpFilename string) error {
+	path := filepath.Join(workDir, tmpFilename)
+	if err := ioutil.WriteFile(path, []byte(content), 0600); err != nil {
```
Comment (thbkrkr) [inline hack/deployer/runner/kyverno/kyverno.go:59]: - #5927:
```suggestion
	if err := os.WriteFile(path, []byte(content), 0600); err != nil {
```

File: hack/deployer/runner/kyverno/kyverno.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package kyverno
+
+import (
+	_ "embed"
+	"fmt"
+	"io/ioutil"
```
Comment (thbkrkr) [inline hack/deployer/runner/kyverno/kyverno.go:10]: - #5927:
```suggestion
```

**Review by thbkrkr — COMMENTED** [review]
Just to confirm, do we want to keep [config/recipes/psp](https://github.com/elastic/cloud-on-k8s/tree/main/config/recipes/psp) for a while, especially since it's referenced in the [docs](https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-recipes.html)?

IIUC we stop to install PSP:
https://github.com/elastic/cloud-on-k8s/blob/721cccfd905f43c345c7957fbfe173ea3ebb6bbf/config/recipes/psp/beats-agent-psp.yaml#L4-L6

But I see ClusterRoles referencing resources that were created by it:
https://github.com/elastic/cloud-on-k8s/blob/0207226ffbb03f8eff9f549ae103e96991a06f41/test/e2e/cmd/run/run.go#L309-L311

https://github.com/elastic/cloud-on-k8s/blob/23275532ef21296a46b9ce35c65a51d890825ed6/config/e2e/roles.yaml#L64-L75

**Comment by barkbay** [comment]
> Just to confirm, do we want to keep [config/recipes/psp](https://github.com/elastic/cloud-on-k8s/tree/main/config/recipes/psp) for a while, especially since it's referenced in the [docs](https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-recipes.html)?

Yes. We however do not need to bind the roles in `config/e2e/roles.yaml` anymore, I'll remove these bindings.

**Comment by barkbay** [comment]
Jenkins test this please

**Comment by barkbay** [comment]
run/e2e-tests

**Comment by barkbay** [comment]
run/e2e-tests

**Comment by barkbay** [comment]
run/e2e-tests

**Pull request merged**